### ISSUE TITLE
Can inference be accelerated and implemented in real time?

### ISSUE THREAD

**Issue by WGS-note** [SWivid/F5-TTS#544]
### Checks

- [X] This template is only for question, not feature requests or bug reports.
- [X] I have thoroughly reviewed the project documentation and read the related paper(s).
- [X] I have searched for existing issues, including closed ones, no similar questions.
- [X] I confirm that I am using English to submit this report in order to facilitate communication.

### Question details

Hi, Can inference be accelerated and implemented in real time?

你好，推理可以加速吗，实现 实时 的效果

**Comment by mame82** [comment]
Saw similar questions here quite often, so I give a few (reusable) answers.

F5-TTS **is** realtime, but it depends on what you try to achieve and how you use a model.

For example on a RTX 4070 Laptop GPU it takes about 1 minute to generate 2 minutes audio, which comes down to a real time factor (RTF) of 0.5 on a moderate GPU.

What F5-TTS is not capable of, is to stream audio back in chunks. In practice this means, if you provide input text like `"Hello, my name is Tom"` the model is not able to stream back small audio chunks like `Hel` ... `lo` ... `my` ... `name` ... `is` ... `Tom`. If this would be the case, it would be possible to play back chunks immediately (low latency output, which is not exactly the same as realtime speed).

What does this mean? Due to its flow based nature, the model creates longer audio output (think of it as whole sentences). Now the issue with the current implementation is, that if your text input exceeds the audio output limit (provided reference audio and generated audio output duration are summed up and mustn't exceeed 30 seconds), audio output generation will be processed in multiple slices. Unfortunately, the implementation of this "batch processing" only returns the concatenated final result (not individual slices, once they are processed). If you provide longish text input, which has to be split into 10 output slices for example and processing of each slice takes 5 seconds, you will receive the result after 50 seconds (which of course makes a long delay, even if the generated output could be two minutes long and processing still was faster than realtime speech).

So if you want faster responses, I suggest to slice your input text before inference. The easiest way to do this (at least in my target language) is to slice on a "per sentence level" or - if slices still are to long - fallback to split text between words or commas to keep input text slices short (avoid slicing by F5-TTS inference code).

This again would allow you, to play back the first sentence once it is processed (lets stay with the example from above, which means after 5 seconds instead of 50 seconds). With a RTF of 0.5 this would mean audio playback of the first sentence takes 10 seconds, in the meantime the next two text slices could be transformed into audio (as processing of a slice takes only 5 seconds). In the end, for the given example, it takes you still 50 seconds to generate all the audio, but playback time is about two minutes and playback starts after 5 seconds (everything assumes RTF 0.5, which depends on your hadware).

So in short words: While the inference implementation supports no streaming, you could do it on your own, but on a per-sentence-level instead of a per-token-level.

**Comment by WGS-note** [comment]
Thank you very much for your reply! I would like to ask one more question: how do I implement the speed of speech?

**Comment by mame82** [comment]
the `infer_process` function takes a `speed` argument which defaults to `1.0`. Setting it to `1.1` for example increases the speed, setting it to `0.9` slows it down (https://github.com/SWivid/F5-TTS/blob/main/src/f5_tts/infer/utils_infer.py#L349).

It should be noted, that the speed setting actually scales the prediction for the output audio duration up or down ... the model basically has to fit in the infered audio, based on the estimated output duration. Say you set speed to `0.5` while your reference audio is "speaking fast". In result the generated audio could have repeated words or other errors (as a twice as long audio output has to be filled, while the reference audio indicates a high word frequency). You have to mess around with speed a bit, iot get used to it. I personally only use it, to allign generated audio with reference audio, in case a speed of `1.0` produces errors with the provided reference audio. If you only want to increase the speed of the speaker (word frequency) just provide a reference audio, where with a faster speking speaker, if that makes sense

**Comment by mame82** [comment]
I agree thatsome more info or "best practices" regarding reference audio should be added to the repo, to clarify some of these things (how speaker speed, pauses between words etc. influence the generated audio). Yet, this could easily be find out with experimentation. Use the gradio inference interface. It provides a "speed" slider iirc and you would also be able to experiment with impact of differences in reference audio, by recording your own voice as reference, right from the gradio interface.

**Comment by WGS-note** [comment]
Thank you for your reply, this was very useful for me! I would also like to ask how to control the speed of speech? Is it based on time delay?

**Comment by SWivid** [comment]
Many thanks @mame82 ~

> how to control the speed of speech

Hi @WGS-note . Our non-autoregressive model is given a total duration to predict the audio (melspectrogram actually) output. So if you have 10 words and order a 10-second for the model to generate, the model itself will decide how to distribute each word's position and duration. If you have the same 10 words but order a 8-second for the model, it is the same but obviously each word's duration is shorter which sounds faster in speaking rate.

For acceleration, some engineering work can help speed up in magnitude (TensorRT, etc.), though we are not familiar with it.
That is also the reason we are open-sourcing, in wish that we could make the TTS technique better with whole community's efforts.

**Comment by WGS-note** [comment]
Thank you very much for your reply! I tried it and it worked very well, because I was using librosa.effects.time_stretch before and the synthesized audio sounded weird, the speed parameter solved my problem very well.

**Comment by mame82** [comment]
The analogy with input genration from above, also helps to understand other discussed aspetcs:

- no streaming of chunks: The model cant hand you back chunks of the "generated image" only the image as a whole (referenced audio + generated audio, based on reference text + generation text)
- if your generation text is too long, say "genrate the right half of the checker board and append more and append even more" it will basically split into multiple "image generations" anf or each "sub generation" the reference image is reused to generate the missing image part with a slice of the generation text (run 1: "generate the right half of the checkerboard", run 2: "and append more", run 3: "and append even more"). The generated parts of run 1 to 3 are then combined and returned **once all generations are runs**
- the analogy to the "step count" of image generation models is "nfe steps", you could lower the to reduce processing time (half nfe steps mean rougly half the processing time) but the result is less defined (bad audio quality) ... going with too many nfe steps on the other hand, brings no noticable improvement of output, while consuming addtional processing time (thats why suggested values for nfe steps is 32 or 64, trade off is speed vs output quality) - depending on the use case, could be fine f.e. dynamic nfe setting (first sentence with low nfe step count, succesive sentences with higher nfe steps, while audio is already playing).

Hope these comments are also helpful for others.

**Comment by DomoMKT** [comment]
you can find the answer to your question on how to accelerate inference and achieve real-time implementation at [www.nemesyslabs.com](https://www.nemesyslabs.com).
